### PR TITLE
Localise module maps

### PR DESCRIPTION
**Pull request by dorchard**
A problem arose in CamFort with the location of declarations being misreported to the wrong file (https://github.com/camfort/camfort/issues/185). It turns out this comes down to the structure of mod file analysis, where all the `use`d modules get their name space inserted into the module map and so it cannot be distinguished whether a variable was declared in this module or an imported one. This PR fixes this by adding information into the `NameType` data structure about whether a variable is `Local` or `Imported`. A module map can then be 'localised' with `localisedModuleMap` which removes from a mod map any imported declarations. This fixes the bug in CamFort. 
An accompanying test was added in `ModFileSpec`

The `genUniqNameToFilenameMap :: ModFiles -> M.Map F.Name String` applies this localisation.

### CODE REVIEW

**Comment by dorchard** [comment]
(note this is rebased off https://github.com/camfort/fortran-src/pull/285)

File: app/Main.hs [hunk 1]
```diff
@@ -50,11 +50,16 @@ import qualified Language.Fortran.Parser.Free.Lexer  as Free
...
+showVersion :: String
+showVersion = "0.16.0"
+
 main :: IO ()
 main = do
   args <- getArgs
   (opts, parsedArgs) <- compileArgs args
   case (parsedArgs, action opts) of
+    (paths, ShowMyVersion) -> do
+      putStrLn $ "fortran-src version: " ++ showVersion
```
Comment (raehik) [inline app/Main.hs:62]: (We should have a release guide that instructs updating the version string shown here. I'll start a wiki page later.)

File: src/Language/Fortran/Util/ModFile.hs [hunk 2]
```diff
@@ -217,6 +217,9 @@ decodeModFiles' = fmap (map snd) . decodeModFiles
 combinedModuleMap :: ModFiles -> FAR.ModuleMap
 combinedModuleMap = M.unions . map mfModuleMap
 
+localisedModuleMap :: FAR.ModuleMap -> FAR.ModuleMap
+localisedModuleMap = M.map (M.filter (not . FA.isImported . snd))
```
Comment (raehik) [inline src/Language/Fortran/Util/ModFile.hs:221]: this threw me for a bit thanks to the deep type synonyms (`ModuleMap` storing a `ModEnv` which itself is actually a `Map`) -- could perhaps clarify what it's doing in a comment

**Review by raehik — APPROVED** [review]
Looks good to me. Some of the pre-existing type synonyms I find hard to follow and might warrant extra documenting, but it's a minor nit.

File: src/Language/Fortran/Util/ModFile.hs (hunk 2 above)
Comment (raehik) [reply src/Language/Fortran/Util/ModFile.hs:221]: that is, I saw `map . M.filter` and my first thought was "isn't this just a filter operation", but we're actually filtering the maps inside our top-level map

Comment (dorchard) [reply src/Language/Fortran/Util/ModFile.hs:223]: Added a comment!

Comment (dorchard) [reply src/Language/Fortran/Util/ModFile.hs:223]: Thanks!
